flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/SimplifyWindowTableFunctionRule.java

```diff
@@ -0,0 +1,355 @@
+
+    public static final RelOptRule WITH_RIGHT_CALC_WINDOW_JOIN =
+            Config.EMPTY
+                    .withDescription("SimplifyWindowTableFunctionRuleWithRightCalcWindowJoin")
+                    .as(Config.class)
+                    .withRightCalcWindowJoin()
+                    .toRule();
+
+    public static final RelOptRule WITH_LEFT_RIGHT_CALC_WINDOW_JOIN =
+            Config.EMPTY
+                    .withDescription("SimplifyWindowTableFunctionRuleWithLeftRightCalcWindowJoin")
+                    .as(Config.class)
+                    .withLeftRightCalcWindowJoin()
+                    .toRule();
+
+    public SimplifyWindowTableFunctionRule(Config config) {
+        super(config);
+    }
+
+    @Override
+    public boolean matches(RelOptRuleCall call) {
+        List<RelNode> rels = call.getRelList();
+        final RelNode node = rels.get(0);
+        if (node instanceof StreamPhysicalWindowRank) {
+            final int windowTVFIdx = rels.size() - 1;
+            final StreamPhysicalWindowTableFunction windowTVF =
+                    (StreamPhysicalWindowTableFunction) rels.get(windowTVFIdx);
+            return needSimplify(windowTVF);
+        } else if (node instanceof StreamPhysicalWindowJoin) {
+            RelNode leftWindowTVFRel;
+            if (call.rule == WITH_WINDOW_JOIN) {
+                leftWindowTVFRel = rels.get(2);
+            } else if (call.rule == WITH_RIGHT_CALC_WINDOW_JOIN) {
+                leftWindowTVFRel = rels.get(2);
+            } else if (call.rule == WITH_LEFT_CALC_WINDOW_JOIN) {
+                leftWindowTVFRel = rels.get(3);
+            } else if (call.rule == WITH_LEFT_RIGHT_CALC_WINDOW_JOIN) {
+                leftWindowTVFRel = rels.get(3);
+            } else {
+                throw new TableException("This should never happen. Please file an issue.");
+            }
+            final StreamPhysicalWindowTableFunction leftWindowTVF =
+                    (StreamPhysicalWindowTableFunction) leftWindowTVFRel;
+            final StreamPhysicalWindowTableFunction rightWindowTVF =
+                    (StreamPhysicalWindowTableFunction) rels.get(rels.size() - 1);
+            return needSimplify(leftWindowTVF) || needSimplify(rightWindowTVF);
+        } else {
+            throw new IllegalStateException(
+                    this.getClass().getName()
+                            + " matches a wrong relation tree: "
+                            + RelOptUtil.toString(node));
+        }
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+        List<RelNode> rels = call.getRelList();
+        final RelNode node = rels.get(0);
+        if (node instanceof StreamPhysicalWindowRank) {
+            RelNode newTree = rebuild(rels);
+            call.transformTo(newTree);
+        } else if (node instanceof StreamPhysicalWindowJoin) {
+            RelNode newLeft;
+            RelNode newRight;
+            if (call.rule == WITH_WINDOW_JOIN) {
```

Comment:
Yes, the rule does not support `WindowAggregate` because for window table function + window Aggregate would be matched by `PullUpWindowTableFunctionIntoWindowAggregateRule`.

Reply:
Because WindowAggregate could handle inputs with `WindowAttachedWindowingStrategy` and `TimeAttributeWindowingStrategy` which means it could handle watermark assigner itself while `WindowRank` and `Windowjoin` can't.